service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,318 @@
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByID(ctx, hashID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) parseBlockBytes(proposerBytes []byte) (*ParsedBlock, error) {
+	errs := wrappers.Errs{}
+
+	proposer, bytes, err := getProposerFromBytes(proposerBytes)
+	if err != nil {
+		return nil, fmt.Errorf("fetching proposer from block bytes errored with %w", err)
+	}
+
+	if p.genesisTimestamp.IsZero() {
+		return nil, errParserUninitialized
+	}
+
+	// Default block time to proposer timestamp if exists, otherwise to genesis block
+	blockTimestamp := new(time.Time)
+
+	if proposer.Timestamp > genesisTimestamp {
+		*blockTimestamp = time.Unix(proposer.Timestamp, 0)
+	} else {
+		*blockTimestamp = time.Unix(genesisTimestamp, 0)
+	}
+
+	var blk blocks.Block
+	ver, err := p.codec.Unmarshal(bytes, &blk)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling block bytes errored with %w", err)
+	}
+	blkID := ids.ID(hashing.ComputeHash256Array(bytes))
+
+	parsedBlock := ParsedBlock{
+		Height:    blk.Height(),
+		BlockID:   blkID,
+		BlockType: fmt.Sprintf("%T", blk),
+		Proposer:  proposer,
+	}
+
+	switch castBlk := blk.(type) {
+	case *blocks.ApricotProposalBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.BlueberryProposalBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.ApricotAtomicBlock:
+		errs.Add(common.InitializeTx(ver, p.codec, castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.ApricotStandardBlock:
+		var blockTxs []*txs.Tx
+
+		for i, tx := range castBlk.Transactions {
+			errs.Add(common.InitializeTx(ver, p.codec, tx))
+			blockTxs = append(blockTxs, castBlk.Transactions[i])
+		}
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = blockTxs
+	case *blocks.BlueberryStandardBlock:
+		var blockTxs []*txs.Tx
+
+		for i, tx := range castBlk.Transactions {
+			errs.Add(common.InitializeTx(ver, p.codec, tx))
+			blockTxs = append(blockTxs, castBlk.Transactions[i])
+		}
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = blockTxs
+	case *blocks.ApricotAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *blocks.BlueberryAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *blocks.BlueberryCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *blocks.ApricotCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	default:
+		errs.Add(fmt.Errorf("no handler exists for block type %T", castBlk))
+	}
+
+	// If the block has an advance time tx, use its timestamp as the block timestamp
```

Comment:
Actually all blueberry blocks have their timestamp explicitly serialized. Moreover in Apricot blocks, only ApricotProposal ones may contain advanceTimeTx (which is then forbidden in Blueberry).